Src/Base/AMReX_CuptiTrace.cpp

```diff
@@ -0,0 +1,294 @@
+#include <AMReX_CuptiTrace.H>
+#ifdef AMREX_USE_CUPTI
+#include <AMReX_Print.H>
+#include <stdio.h>
+#include <map>
+#include <cuda.h>
+#include <cupti.h>
+
+
+namespace amrex {
+
+std::vector<CUpti_Activity_Userdata*> activityRecordUserdata;
+
+
+void CUPTIAPI
+bfrRequestCallback (uint8_t* *bfr, size_t* size, size_t* maxNumRecords) noexcept
+{
+  // Allocate a buffer for use by CUPTI; activity records are stored in the buffer
+  uint8_t* buffer = (uint8_t*) malloc(BFR_SIZE + ALIGNMENT);
+  if (buffer == NULL) {
+    amrex::Print() << "Error: CUPTI requested a buffer but memory cannot be allocated\n";
+    exit(EXIT_FAILURE);
+  }
+
+  // Return buffer and buffer size
+  *bfr = ALIGN_BFR(buffer, ALIGNMENT);
+  *size = BFR_SIZE;
+
+  // Controls maximum number of records to be placed in buffer; setting to zero
+  // fills the buffer with as many records as possible, setting to n > 0 fills
+  // with n records before buffer is returned
+  *maxNumRecords = 0;
+}
+
+void CUPTIAPI
+bfrCompleteCallback (CUcontext ctx, uint32_t streamId, uint8_t* bfr,
+		     size_t size, size_t validSize) noexcept
+{ 
+  CUptiResult status;
+  CUpti_Activity* record = NULL;
+
+  if (validSize > 0) {
+    do {
+      status = cuptiActivityGetNextRecord(bfr, validSize, &record);
+      if (status == CUPTI_SUCCESS) {
+	CUpti_Activity_Userdata* recordUserData = new CUpti_Activity_Userdata();
```

Comment:
Where is this `delete`d?

Reply:
I had a function `void CuptiTrace::cleanup () noexcept` to either call manually or that is called by the destructor of a `CuptiTrace` object, which clears the `recordUserData`s.  is this not a safe way to do it, since there is no guarantee the `cleanup` function is called?

Reply:
Yes, it's safe.  For example, you call in `cuptiTraceStart`, `std::vector::clear` is called.  That will not release the memory.  I actually missed the `delete[]` in `cleanup` because I was searching for `delete ` with a space after `delete`.  The reason I searched for deleted was the object was made with `new`, not `new[]`.  Note it is an *error* to call `delete[]` on an object made with `new`.  One can only call `delete[]` on objects made with `new[]`.  My suggestion is to make it `std::vector<std::unique_ptr<CUpti_Activity_UserData>>`.

Reply:
Thanks for explaining, I have switched to `std::vector<std::unique_ptr<CUpti_Activity_UserData>>` as you suggested.  I hope this and similar issues are now avoided, please let me know if it looks OK.